### Importing Libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

### Importing Data

In [2]:
# import products.csv as df_prods from the original data folder
path = r'/Users/profkdfischer/Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [3]:
# check that df_prods looks okay
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [4]:
# check the number of rows in df_prods
df_prods.shape

(49693, 5)

In [5]:
# import orders-wrangled.csv as df_ords from the prepared data folder
path = r'/Users/profkdfischer/Instacart Basket Analysis'
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Exercise 4.4_orders_wrangled.csv'),  index_col = False)

In [6]:
# check that df_ords looks ok
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   orders_time  days_since_prior_order  
0            8                     NaN  
1            7                    15.0  
2           12                    21.0  
3            7                    29.0  
4           15                    28.0

In [7]:
# check the number of rows in df_ords
df_ords.shape

(3421083, 8)

## Exercise 4.5 - working through data consistency examples

#### Mixed Datatype Columns:

In [8]:
# make a tester df for testing how to deal with mixed datatype columns
df_test = pd.DataFrame()
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
# check it
df_test

mix
0     a
1     b
2     1
3  True

In [10]:
# check for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)
  else:
    print('consistent')

mix


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/1311243746.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [11]:
# fix the mixed data types by changing all to str
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
# check if the "mix" column has been fixed and no longer has mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)
  else:
    print('consistent')

consistent


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/778994023.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


#### Missing Values:

In [13]:
# check for missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
# make a subset of these missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [15]:
# check it
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
# check how many rows in original df_prods
df_prods.shape

(49693, 5)

In [17]:
# make a clean subset of df_prods without the NaN values in 'product_name'
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
# check how many rows in the new clean df without NaN values
df_prods_clean.shape

(49677, 5)

There were 16 rows with NaN values. Therefore, our new, clean product df has 16 fewer rows than the original products df.

#### Duplicate values:

In [19]:
# find rows with duplicate values
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
# look at these duplicate rows
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
# drop the duplicates from df_prods_clean
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
# check how many rows in the new df without the duplicate rows
df_prods_clean_no_dups.shape

(49672, 5)

In [23]:
# list all columns
df_prods_clean_no_dups.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

There were 5 duplicate rows. 
Therefore, our new cleaner df, df_prods_clean_no_dups, has 5 fewer rows than the previous version

In [24]:
# export this cleaned version of products (with missing values and duplicates removed)
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## Exercise 4.5 - Tasks

#### Q2: Run the df.describe() function on your df_ords dataframe...share in a markdown cell whether anything about the data looks off or should be investigated further. 

In [25]:
# check df_ords descriptive stats
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week   orders_time  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

Q2 answer: 
In the "days_sinc_prior_order" column, we see that the mean value is 11.1 days, while the median value is 7 days. This means that there some outlier values that are pushing the mean higher than the median, or in other words, some customers that are taking a lot more time to fill a second order than the median number. We should try to figure out what is stopping those customers from re-ordering on time.

#### Q3and Q4: Check for mixed-type data in your df_ords dataframe. If you find mixed-type data, fix it.

In [26]:
# check for mixed data types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
  else:
    print('consistent')

/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/965142442.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


consistent


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/965142442.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


consistent


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/965142442.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


consistent


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/965142442.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


consistent


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/965142442.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


consistent


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/965142442.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


consistent


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/965142442.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


consistent


/var/folders/s0/b631rndx6sv1py5f9s6wc7y00000gn/T/ipykernel_3189/965142442.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)


consistent


Q3 and Q4 answers: 
I did not find any mixed datatype columns in df_ords. No fixing required.

#### Q5: Run a check for missing values in your df_ords dataframe. 
#### Q6: Address the missing values using an appropriate method. 

In [27]:
# check for missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
orders_time                    0
days_since_prior_order    206209
dtype: int64

Q5 answer: The only missing values appear in the "days_since_prior_order" column. These must represent customers who only ordered once - because there was not a second order yet, the calculation returned a "NaN". 

Q6 answer: There are 3 methods for dealing with missing values: (1) create a new variable that acts like a flag (2) impute (3) filter out. I would choose method (1) here and create a new variable. The reason for this is that these values are important since they identify customers who have only ordered once from us. I would create a new column and call the new variables "new customers" (the 206209 missing values).

In [28]:
# address the missing values in df_ords by creating a new column to identify the NaN values as new customers
df_ords_clean = df_ords
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [29]:
# check the top 5 rows of df_ords_clean
df_ords_clean

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  orders_time  days_since_prior_order  new_customer  
0                         2            8                     NaN          True  
1                         3            7                    15.0         False  
2                         3           12                    21.0         False  
3                         4            7                    29.0         False  
4                         4           15                    28.0         False  
...                     ...          ...                     ...           ...  
3421078                   5           18                    29.0         False  
3421079                   4           10                    30.0         False  
3421080                   1           12                    18.0         False  
3421081                   1           12                     7.0         False  
3421082                   6           14                    30.0         False  

[3421083 rows x 9 columns]

In [30]:
# check the number of rows in df_ords_clean
df_ords_clean.shape

(3421083, 9)

#### Q7: Run a check for duplicate values in your df_ords data. 
#### Q8: Address the duplicates using an appropriate method.

In [31]:
# find rows with duplicate values
df_dups_ords_clean = df_ords_clean[df_ords.duplicated()]

In [32]:
# check the duplicate values df
df_dups_ords_clean

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, orders_time, days_since_prior_order, new_customer]
Index: []

Q7 answer: There are no duplicate values in df_ords.
Q8 answer: No action required.

#### Q9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder.

In [33]:
# export the cleaned version of orders

df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_consistent.csv'))

In [34]:
# export the cleaned version of products

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_consistent.csv'))